tests/test_redirects.py

```diff
@@ -0,0 +1,211 @@
+
+        self.assertEqual(result[0]["newUrl"], "http://localhost/test-page-url/")
+        self.assertEqual(result[1]["newUrl"], "http://localhost/test-page-url/")
+        self.assertEqual(result[2]["newUrl"], "http://test/just-link")
+
+    def test_no_new_url_query(self):
+        """
+        Test that a redirect can be created without a `redirect_link` or
+        `redirect_page` and queried. This can be used for producing a HTTP 410
+        for any pages/URLs that are no longer in existence.
+        """
+        self.redirect = RedirectFactory(
+            redirect_link="",
+            redirect_page=None,
+        )
+
+        query = """
+        {
+            redirects {
+                newUrl
+            }
+        }
+        """
+
+        result = self.client.execute(query)["data"]["redirects"][0]["newUrl"]
+
+        self.assertEqual(result, None)
+
+    def test_specified_site_url(self):
+        """
+        Test that a redirect with a specified site returns the correct `old_url`.
+        """
+        self.redirect = RedirectFactory(
+            old_path="old-path",
+            site=SiteFactory(
+                hostname="test-site",
+                port=81,
+            ),
+        )
+        self.redirect = RedirectFactory(
+            old_path="old-path",
+            site=SiteFactory(
+                hostname="test-site-default-port",
+                port=80,
+            ),
+        )
+        self.redirect = RedirectFactory(
+            old_path="old-path",
+            site=SiteFactory(
+                hostname="test-site-secure",
+                port=443,
+            ),
+        )
+
+        query = """
+        {
+            redirects {
+                oldUrl
+            }
+        }
+        """
+
+        result = self.client.execute(query)["data"]["redirects"]
+
+        self.assertEqual(result[0]["oldUrl"], "http://test-site:81/old-path")
+        self.assertEqual(result[1]["oldUrl"], "http://test-site-default-port/old-path")
+        self.assertEqual(result[2]["oldUrl"], "https://test-site-secure/old-path")
+
+    def test_all_sites_url(self):
+        """
+        Test that when no site is specified on a redirect, that redirect is
+        shown for each existing site.
+        """
+        self.site1 = SiteFactory(hostname="test-site", port=81)
+        self.site2 = SiteFactory(hostname="another-test-site", port=82)
+
+        self.redirect = RedirectFactory(
+            old_path="old-path",
+            site=None,
+        )
+
+        query = """
+        {
+            redirects {
+                oldUrl
+            }
+        }
+        """
+
+        result = self.client.execute(query)["data"]["redirects"]
+
+        self.assertEqual(result[0]["oldUrl"], "http://another-test-site:82/old-path")
+        self.assertEqual(result[1]["oldUrl"], "http://localhost/old-path")
+        self.assertEqual(result[2]["oldUrl"], "http://test-site:81/old-path")
```

Comment:
There are some additional checks that would be beneficial to add to this test to ensure the data is in the correct format before comparing values.

Here are the comments I would find helpful when reading the code:

```suggestion
    def test_all_sites_url(self):
        """
        Test that when no site is specified on a redirect, that redirect is
        shown for each existing site.
        """

        # Create two additional sites (on top of the default `Localhost` one).
        SiteFactory(hostname="test-site", port=81)
        SiteFactory(hostname="another-test-site", port=82)

        # Create a `Redirect` without a `site`.
        RedirectFactory(
            old_path="old-path",
            site=None,
        )

        query = """
        {
            redirects {
                oldUrl
            }
        }
        """

        result = self.client.execute(query)["data"]["redirects"]

        # Sanity check to ensure there are three `Site`s.
        self.assertEqual(Site.objects.count(), 3)
        
        # Sanity check to ensure only one `Redirect` is in the database.
        self.assertEqual(Redirect.objects.count(), 1)

        # Ensure there are three `Redirect`s returned despite only 1 object
        # in the database.
        self.assertEqual(len(result), 3)

        self.assertEqual(result[0]["oldUrl"], "http://another-test-site:82/old-path")
        self.assertEqual(result[1]["oldUrl"], "http://localhost/old-path")
        self.assertEqual(result[2]["oldUrl"], "http://test-site:81/old-path")
```

Can you experiment with this and see if any other tests would benefit from additional checks (like checking the number of results)?

Reply:
Thanks! Improved documentation of `test_redirects.py` in 40eeb98e50bd5cf9895d758fbd761f3f2c474567